### PR TITLE
[react] making sure ReactDOMServer always uses ServerReactRootIndex, …

### PR DESCRIPTION
**Pull request by fabiomcosta**
When ReactDOMServer runs on any environment with a DOM (browser) it generates ids `ClientReactRootIndex` which is inconsistent with what happens on the client. This is also a problem if you're using the result of `ReactDOMServer.renderToString` to cache components and having to render use that on different page renders.

This diff makes `ReactDOMServer.renderToString` always use `ServerReactRootIndex. createReactRootIndex` and `ReactMount.render` always use `ClientReactRootIndex. createReactRootIndex`.

### CODE REVIEW

**Comment by zpao** [comment]
cc @spicyj. Whatever happens here I don't think we want to take this into 0.14 - definitely sounds like a behavioral change from RC.

**Comment by spicyj** [comment]
Yeah, probably not for 0.14. @fabiomcosta just wants it internally.

Will take a closer look tomorrow, but you can (and should) get rid of the injection now and require the two modules directly.

**Comment by fabiomcosta** [comment]
@spicyj sounds good to me! I'll update the diff accordingly, thank you!

**Comment by fabiomcosta** [comment]
Made sure there are no references to `ReactRootIndex`.
This has to be documented as a change to the next version, let me know if I can help with that.

File: src/renderers/dom/client/__tests__/ReactRenderDocument-test.js [hunk 1]
```diff
@@ -213,8 +228,8 @@ describe('rendering React components at document', function() {
       'quirks by rendering at the document root. You should look for ' +
       'environment dependent code in your components and ensure ' +
       'the props are the same client and server side:\n' +
-      ' (client) data-reactid=".0.1">Hello world</body></\n' +
-      ' (server) data-reactid=".0.1">Goodbye world</body>'
+      ' (client) ata-reactid=".-1.1">Hello world</body></\n' +
```
Comment (spicyj) [inline src/renderers/dom/client/__tests__/ReactRenderDocument-test.js]: …what happened to the `d`?

File: src/renderers/dom/client/__tests__/ReactRenderDocument-test.js [hunk 2]
```diff
@@ -30,7 +30,22 @@ var UNMOUNT_INVARIANT_MESSAGE =
 
 describe('rendering React components at document', function() {
   beforeEach(function() {
-    require('mock-modules').dumpCache();
+    require('mock-modules')
+      .dumpCache()
+      .mock('ClientReactRootIndex')
```
Comment (spicyj) [inline src/renderers/dom/client/__tests__/ReactRenderDocument-test.js]: Normally we put the mock calls at the top level – does that work?

**Comment by spicyj** [comment]
I think this change is minor enough that it probably doesn't need to be documented, but we shouldn't take it before 0.14. (If you need it at FB before then we can figure something out.)

File: src/renderers/dom/client/__tests__/ReactRenderDocument-test.js (hunk 1 above)
Comment (fabiomcosta) [inline src/renderers/dom/client/__tests__/ReactRenderDocument-test.js]: It gets a fixed number of characters, with the addition of the minus sign it got removed from the output

File: src/renderers/dom/client/__tests__/ReactRenderDocument-test.js (hunk 2 above)
Comment (fabiomcosta) [inline src/renderers/dom/client/__tests__/ReactRenderDocument-test.js]: I can move it up, but this works just fine.

**Comment by fabiomcosta** [comment]
Should we also rename things?
`ClientReactRootIndex` -> `SequentialReactRootIndex`
`ServerReactRootIndex` -> `RandomReactRootIndex`

File: src/renderers/dom/client/__tests__/ReactRenderDocument-test.js (hunk 1 above)
Comment (spicyj) [inline src/renderers/dom/client/__tests__/ReactRenderDocument-test.js]: Oh, right.

File: src/renderers/dom/client/__tests__/ReactRenderDocument-test.js (hunk 2 above)
Comment (spicyj) [inline src/renderers/dom/client/__tests__/ReactRenderDocument-test.js]: Please do so that we're consistent.

**Comment by spicyj** [comment]
Names are fine as-is.

**Comment by spicyj** [comment]
Can you squash your commits together?

**Comment by fabiomcosta** [comment]
Sure

**Comment by fabiomcosta** [comment]
I'm working on fixing this.

**Comment by sebmarkbage** [comment]
We're already using this so I guess it is accepted in some form.

**Comment by fabiomcosta** [comment]
Let me know if any action is necessary from me

**Pull request merged**